src/mappings/Cache.ts

```diff
@@ -0,0 +1,104 @@
+    const blockId = blockNumber(event)
+    let cachedBlock = rewardDestinationByAddress[blockId]
+    
+    if (cachedBlock !== undefined) {
+        return cachedBlock[accountAddress]
+    } else {
+        rewardDestinationByAddress = {}
+        
+        let method = event.event.method
+        let section = event.event.section
+
+        const allAccountsInBlock = event.block.events
+            .filter(blockEvent => { 
+                return blockEvent.event.method == method && blockEvent.event.section == section
+            })
+            .map(event => { 
+                let {event: {data: [accountId, ]}} = event
+                return accountId
+            });
+
+        const payees = await api.query.staking.payee.multi(allAccountsInBlock);
+        const rewardDestinations = payees.map(payee => { return payee as RewardDestination });
+        
+        let destinationByAddress: {[address: string]: RewardDestination} = {}
+        
+        // something went wrong, so just query for single accountAddress
+        if (rewardDestinations.length !== allAccountsInBlock.length) {
+            const payee = await api.query.staking.payee(accountAddress);
+            destinationByAddress[accountAddress] = payee;
+            rewardDestinationByAddress[blockId] = destinationByAddress
+            return payee
+        }
+        allAccountsInBlock.forEach((account, index) => { 
+            let accountAddress = account.toString()
+            let rewardDestination = rewardDestinations[index]
+            destinationByAddress[accountAddress] = rewardDestination
+        })
+        rewardDestinationByAddress[blockId] = destinationByAddress
+        return destinationByAddress[accountAddress]
+    }
+}
+
+export async function cachedController(accountAddress: string, event: SubstrateEvent): Promise<string> {
+    const blockId = blockNumber(event)
+    let cachedBlock = controllersByStash[blockId]
+    
+    if (cachedBlock !== undefined) {
+        return cachedBlock[accountAddress]
+    } else {
+        controllersByStash = {}
+        
+        let method = event.event.method
+        let section = event.event.section
+
+        const allAccountsInBlock = event.block.events
+            .filter(blockEvent => { 
+                return blockEvent.event.method == method && blockEvent.event.section == section
+            })
+            .map(event => { 
+                let {event: {data: [accountId, ]}} = event
+                return accountId
+            });
+
+        var controllerNeedAccounts: AccountId[] = []
+
+        for (let accountId of allAccountsInBlock) {
+            const rewardDestination = await cachedRewardDestination(accountId.toString(), event)
+
+            if (rewardDestination.isController) {
+                controllerNeedAccounts.push(accountId as AccountId)
+            }
+        }
+
+        const bonded = await api.query.staking.bonded.multi(controllerNeedAccounts);
```

Comment:
What if `controllerNeedAccounts` is empty?

Reply:
There are no use cases where it is possible but let's handle